apiserver: wire OpenAPI into sample-apiserver

sttts opened the pull request:
This PR add OpenAPI generator to code-generator scripts, and generated and wires the spec in sample-apiserver.

Note: this intentionally checks in the codegen and sample-apiserver spec because the former is a good place to track changes to tooling and the later because an (compilable) example needs it to be checked in. The reason we don't do that at the top-level is because API PRs often conflict at the spec file. This is not the case for codegen and sample-apiserver.

This is also prerequisite for proper separation of the delegated API server specs in https://github.com/kubernetes/kubernetes/pull/74133, which will also clean up technical debt we took when adding CRD openapi publishing, to create the apiextensions-apiserver spec.

```release-note
The sample-apiserver gains support for OpenAPI v2 spec serving at `/openapi/v2`.

The `generate-internal-groups.sh` script in k8s.io/code-generator will generate OpenAPI definitions by default in `pkg/generated/openapi`. Additional API group dependencies can be added via `OPENAPI_EXTRA_PACKAGES=<group>/<version> <group2>/<version2>...`.
```

Commit:
code-generator: call openapi-gen from generate-internal-groups.sh

OpenAPI schema is a requirement for API Servers to be used with API Aggregation
and for kubectl to talk to them.

sttts committed:
sample-apiserver: add openapi spec

sttts commented:
/assign @cblecker 

For the build/ change.

sttts commented:
/assign @nikhita 

for the sample-apiserver and code-gen change.

yue9944882 reviewed (approved):
LGTM, i didn't see any openapi marker tags in the example models tho...

sttts commented:
@yue9944882 because we had them already there. But dysfunctional.

cblecker commented:
Looks like both the cross build failures and the verify failures are legitimate

sttts committed:
apiextensions: exclude newly added openapi from hack/update-codegen.sh

This will we readded in a follow-up.

sttts committed:
Makefile.generated_files: build code-generator+sample-apiserver OpenAPI schemas

yliaog commented:
/cc @roycaihw

sttts committed:
apiextensions: update OpenAPI spec with v1beta1 meta types

sttts committed:
openapi: commit low-change code-gen+sample-apiserver specs

The former is used as a smoke test to notice changes. The latter is supposed to be published and therefore must be commited. Both do not change much.

sttts commented:
/retest

liggitt commented on build/root/.kazelcfg.json:
```diff
@@ -3,7 +3,8 @@
 		"/_",
-		"^third_party/etcd.*"
+		"^third_party/etcd.*",
+		"k8s.io/sample-apiserver/pkg/apis/.*"
```
no idea what this is doing... surprised we needed this added here if `staging/src/k8s.io/apiextensions-apiserver/pkg/generated/openapi/zz_generated.openapi.go` didn't require it

sttts replied:
Oh, I think we can remove it now that Makefile.generated_files is extended.

sttts replied:
fixed

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/hack/update-codegen.sh:
```diff
@@ -27,7 +27,7 @@ CODEGEN_PKG=${CODEGEN_PKG:-$(cd "${SCRIPT_ROOT}"; ls -d -1 ./vendor/k8s.io/code-
 CLIENTSET_NAME_VERSIONED=clientset \
 CLIENTSET_NAME_INTERNAL=internalclientset \
-"${CODEGEN_PKG}/generate-internal-groups.sh" all \
+"${CODEGEN_PKG}/generate-internal-groups.sh" deepcopy,client,lister,informer,conversion \
```
why remove this?

sttts replied:
We add openapi to "all", but are not ready for apiextensions-apiserver yet. I have a follow-up to readd it.

liggitt replied:
probably needs to be in the release notes for other consumers of code-gen?

sttts replied:
added

liggitt commented on hack/verify-generated-files-remake.sh:
```diff
@@ -291,11 +291,14 @@ fi
 assert_clean
 
-touch "staging/src/k8s.io/api/core/v1/types.go"
+touch "staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go"
```
should we touch both files?

sttts replied:
Actually I want both tests: apimachinery leads to all specs regenerated, core/v1 only to an updated main spec.

sttts replied:
done

sttts commented:
/retest

sttts commented:
@cblecker @liggitt updated and addressed comments. Ptal.

liggitt commented on staging/src/k8s.io/code-generator/generate-internal-groups.sh:
```diff
@@ -48,6 +48,18 @@ GROUPS_WITH_VERSIONS="$5"
+    done
+    echo k8s.io/kube-openapi
+)
+go install "${KUBE_OPENAPI_PKG}"/cmd/openapi-gen
```
does code-generator have consumers via the published staging repo? what does this addition do for those consumers?

sttts replied:
let me check. I wonder whether we shouldn't put a wrapper into k8s.io/code-generator to reuse the path magic.

sttts replied:
done

liggitt commented on hack/verify-generated-files-remake.sh:
```diff
@@ -56,7 +56,7 @@ function newer() {
             echo "${F}"
         fi
-    done
+    done | sort
```
some other places add `LC_ALL=C` to avoid locale-specific sort differences... does that matter for these?

sttts replied:
those are package names which probably do not fall under locale-specific sorting. But adding LC_ALL=C certainly does not harm.

sttts replied:
done

liggitt commented:
a couple minor comments, but otherwise sample-apiserver changes lgtm

I verified the `make generated_files` rebuilds the generated openapi files, but I'm not sure I'm qualified to review the make file changed directly.

liggitt commented:
needs update-vendor.sh to recompute dependencies of codegen

lgtm otherwise. would like ack on the build changes from @cblecker

sttts committed:
Update vendor

sttts commented:
@liggitt updated vendor. @cblecker ptal

spiffxp commented on hack/verify-generated-files-remake.sh:
```diff
@@ -286,12 +286,32 @@ if [[ -n "${X}" ]]; then
 assert_clean
 
-touch "staging/src/k8s.io/api/core/v1/types.go"
+touch "staging/src/k8s.io/api/apps/v1/types.go"
```
why the change from core to apps here?

sttts replied:
Because core/v1 is a dependency for autoscaling/v1, which is needed by apiextensions-apiserver (in a follow-up). apps/v1 is only needed by kube-apiserver, and therefore is a better choice here.

spiffxp commented on staging/src/k8s.io/code-generator/generate-internal-groups.sh:
```diff
@@ -108,3 +109,14 @@ if [ "${GENS}" = "all" ] || grep -qw "informer" <<<"${GENS}"; then
+
+if [ "${GENS}" = "all" ] || grep -qw "openapi" <<<"${GENS}"; then
+  echo "Generating OpenAPI definitions for ${GROUPS_WITH_VERSIONS} at ${OUTPUT_PKG}/openapi"
+  declare -a OPENAPI_EXTRA_PACKAGES
```
why is this needed? we don't appear to be populating it at all

(also arrays in bash is a sign to start considering something other than bash)

sttts replied:
because it will be needed for apiextensions-apiserver (follow-up PR) which needs autoscaling/v1 as additional group.

sttts replied:
> also arrays in bash is a sign to start considering something other than bash

It's our choice in k8s.io/code-generator for project not using kubebuilder.

The `declare -a` makes sure that `${OPENAPI_EXTRA_PACKAGES[@]}` in the next line does not fail when the variable is empty.

spiffxp replied:
Yeah, I understand this is the language you're locked into, and it's a sane/clean declaration given that constraint. I just wanted to make sure this wasn't something being left around just incase, follow on makes sense.

spiffxp reviewed (commented):
Didn't have time to wade through the Makefile, but some questions about the bash surrounding it

fejta reviewed (commented):
/approve

spiffxp commented:
/lgtm
That Makefile is pretty eye-watering stuff, but the changes are sanely written for what it is.  Thanks for the consistent ordering and use of `prefix` and the answers to my bash questions.  I can't speak to the dependencies and such being correct, but I'll trust that to the /approve

cblecker commented:
Reviewed and mostly looks sane. One open question, otherwise LGTM.

sttts commented on build/root/Makefile.generated_files:
```diff
@@ -408,81 +408,97 @@ $(CONVERSION_GEN): $(k8s.io/kubernetes/vendor/k8s.io/code-generator/cmd/conversi
 # updated API violations to the known API violation exceptions list.
 ifeq ($(UPDATE_API_KNOWN_VIOLATIONS),true)
-    REPORT_FILENAME:=$(KNOWN_VIOLATION_FILENAME)
+    REPORT_FILENAME:=$(KUBE_KNOWN_VIOLATION_FILENAME)
```
Fixed.

This code path was even broken before this PR because we forgot the dependency from the violation file (which is touched) to the output. I fixed that. And I added a test case.

liggitt commented:
/lgtm
/approve

sttts commented:
Rebased on changed ObjectMeta comment.

spiffxp commented:
/lgtm
